arangod/Aql/OptimizerRules.cpp

```diff
@@ -4057,33 +4155,62 @@ class RestrictToSingleShardChecker final : public WalkerWorker<ExecutionNode> {
       case EN::TRAVERSAL:
       case EN::SHORTEST_PATH: {
         _stop = true;
-        return true; // abort enumerating, we are done already!
+        return true;  // abort enumerating, we are done already!
+      }
+
+      case EN::FILTER: {
+        auto collectionVariables =
+            ExecutionNode::castTo<FilterNode const*>(en)->collectionVariables();
+        for (auto pair : collectionVariables) {
+          aql::Collection const* collection = pair.second;
+          aql::Variable const* collectionVariable = pair.first;
+          std::string shardId =
+              getSingleShardId(_plan, en, collection, collectionVariable);
+          if (shardId.empty() && _shardsUsed[collection].empty()) {
+            _shardsUsed[collection].emplace("all");
+          } else if (!shardId.empty()) {
+            if (1 == _shardsUsed[collection].size() &&
+                "all" == *_shardsUsed[collection].begin()) {
+              _shardsUsed[collection].clear();
+            }
+            _shardsUsed[collection].emplace(shardId);
+          }
+        }
+        break;
       }
 
       case EN::INDEX: {
         // track usage of the collection
-        auto collection = ExecutionNode::castTo<IndexNode const*>(en)->collection();
+        auto collection =
+            ExecutionNode::castTo<IndexNode const*>(en)->collection();
         std::string shardId = getSingleShardId(_plan, en, collection);
-        if (shardId.empty()) {
-          _shardsUsed[collection].clear();
+        if (shardId.empty() && _shardsUsed[collection].empty()) {
           _shardsUsed[collection].emplace("all");
-        } else {
+        } else if (!shardId.empty()) {
+          if (1 == _shardsUsed[collection].size() &&
+              "all" == *_shardsUsed[collection].begin()) {
+            _shardsUsed[collection].clear();
+          }
           _shardsUsed[collection].emplace(shardId);
         }
         break;
       }
 
       case EN::ENUMERATE_COLLECTION: {
         // track usage of the collection
-        auto collection = ExecutionNode::castTo<EnumerateCollectionNode const*>(en)->collection();
-        _shardsUsed[collection].clear();
-        _shardsUsed[collection].emplace("all");
+        auto collection =
+            ExecutionNode::castTo<EnumerateCollectionNode const*>(en)
+                ->collection();
+        if (_shardsUsed[collection].empty()) {
+          _shardsUsed[collection].emplace("all");
+        }
         break;
       }
 
       case EN::UPSERT: {
         // track usage of the collection
-        auto collection = ExecutionNode::castTo<ModificationNode const*>(en)->collection();
+        auto collection =
+            ExecutionNode::castTo<ModificationNode const*>(en)->collection();
```

Comment:
An UPSERT on a collection should previously have turned off the single-shard-optimization for a collection.
I think it has to stay like this, as we do not support single-shard-optimization for UPSERT queries yet AFAIK.